### PR TITLE
Us59

### LINKED ISSUE #59: DocumentReadStream
As a Smalltalk programmer, I want to be able to easily iterate over a RichTextDocument (see #57 ), in order to easier create scripts involving them.

**Acceptance Criteria**
- [x] A ReadStream for DocumentObjects extists
- [x] When opening the ReadStream on a RichTextDocument and calling next, it returns the next section of Text and the list of stop conditions as a Tuple (or Association)
- [x] Possible Stop conditions are: (they should always be ordered according to this list)
  - [x] End of a structure
  - [x] End of a paragraph
  - [x] A new line
  - [x] Beginning of a structure
Stop conditions containing a structure must include what structure changed and how it changed

*First step toward a unified scripting API.
Should allow for easy implementation of* #39

Estimated effort: 8

### PR DESCRIPTION
**Pull request by oleschl**

### CODE REVIEW

File: packages/RichTextEditing-Core.package/DocumentReadStream.class/class/on..st [hunk 1]
```diff
@@ -0,0 +1,4 @@
+as yet unclassified
```
Comment (frcroth) [inline packages/RichTextEditing-Core.package/DocumentReadStream.class/class/on..st]: classify

File: packages/RichTextEditing-Core.package/DocumentReadStream.class/instance/paragraphCharacter.st [hunk 2]
```diff
@@ -0,0 +1,4 @@
+as yet unclassified
```
Comment (frcroth) [inline packages/RichTextEditing-Core.package/DocumentReadStream.class/instance/paragraphCharacter.st]: classify

File: packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/boldTextStructure.st [hunk 3]
```diff
@@ -0,0 +1,4 @@
+as yet unclassified
```
Comment (frcroth) [inline packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/boldTextStructure.st]: constants

Comment (frcroth) [reply packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/boldTextStructure.st]: These structures should probably be defined in the superclass

File: packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/document.st [hunk 4]
```diff
@@ -0,0 +1,4 @@
+as yet unclassified
```
Comment (frcroth) [inline packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/document.st]: classify

File: packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/isBeginningOfStructure..st [hunk 5]
```diff
@@ -0,0 +1,4 @@
+as yet unclassified
```
Comment (frcroth) [inline packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/isBeginningOfStructure..st]: classify,

also class should be renamed to something like DocumentReadStreamTests

File: packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/isEndOfStructure..st [hunk 6]
```diff
@@ -0,0 +1,6 @@
+as yet unclassified
```
Comment (frcroth) [inline packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/isEndOfStructure..st]: classify

File: packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/isEndOfStructure..st [hunk 7]
```diff
@@ -0,0 +1,6 @@
...
+isEndOfStructure: listElement
+
+	^ listElement first = #endOfStructure
+
+
```
Comment (frcroth) [inline packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/isEndOfStructure..st]: newlines

File: packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/isParagraph..st [hunk 8]
```diff
@@ -0,0 +1,4 @@
+as yet unclassified
```
Comment (frcroth) [inline packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/isParagraph..st]: c

File: packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/italicTextStructure.st [hunk 9]
```diff
@@ -0,0 +1,4 @@
+as yet unclassified
```
Comment (frcroth) [inline packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/italicTextStructure.st]: c

File: packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/paragraphCharacter.st [hunk 10]
```diff
@@ -0,0 +1,4 @@
+as yet unclassified
```
Comment (frcroth) [inline packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/paragraphCharacter.st]: c

File: packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/testUS59EndOfStructureAndBeginOfStructureIncludeStructure.st [hunk 11]
```diff
@@ -0,0 +1,9 @@
+as yet unclassified
```
Comment (frcroth) [inline packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/testUS59EndOfStructureAndBeginOfStructureIncludeStructure.st]: acceptance-tests

File: packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/testUS59NextReturnsNextSectionOfTextAndListOfStopConditions.st [hunk 12]
```diff
@@ -0,0 +1,9 @@
+as yet unclassified
```
Comment (frcroth) [inline packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/testUS59NextReturnsNextSectionOfTextAndListOfStopConditions.st]: acceptance-tests

File: packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/testUS59ReadStreamExists.st [hunk 13]
```diff
@@ -0,0 +1,4 @@
+as yet unclassified
```
Comment (frcroth) [inline packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/testUS59ReadStreamExists.st]: acceptance-tests

File: packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/testUS59ReadStreamExists.st [hunk 14]
```diff
@@ -0,0 +1,4 @@
...
+testUS59ReadStreamExists
+
+	self assert: (Smalltalk hasClassNamed: #DocumentReadStream)
```
Comment (frcroth) [inline packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/testUS59ReadStreamExists.st]: weak test

File: packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/testUS59StopConditionsAreInCorrectOrder.st [hunk 15]
```diff
@@ -0,0 +1,10 @@
+as yet unclassified
```
Comment (frcroth) [inline packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/testUS59StopConditionsAreInCorrectOrder.st]: acceptance-tests

File: packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/textWithAttributes.st [hunk 16]
```diff
@@ -0,0 +1,8 @@
+as yet unclassified
```
Comment (frcroth) [inline packages/RichTextEditing-Tests.package/RichTextDocumentReadStream.class/instance/textWithAttributes.st]: constants/ utilities

**Review by frcroth — CHANGES_REQUESTED** [review]

File: packages/RichTextEditing-Core.package/DocumentReadStream.class/instance/collection..st [hunk 17]
```diff
@@ -0,0 +1,4 @@
+accessing
+collection: aCollection
```
Comment (konradh) [inline packages/RichTextEditing-Core.package/DocumentReadStream.class/instance/collection..st]: Maybe we could name the accessors `text:` and `text ` to specify that the variable contains the text of the document (still using the instance var `collection`).

File: packages/RichTextEditing-Core.package/DocumentReadStream.class/instance/endOfStructure.st [hunk 18]
```diff
@@ -0,0 +1,11 @@
+stop conditions
+endOfStructure
+	| currentStructures nextStructures diff |
+	
+	self position + 1 > self readLimit ifTrue: [^ nil].
+	currentStructures := self collection attributesAt: self position.
+	nextStructures := self collection attributesAt: self position + 1.
+	diff := currentStructures difference: nextStructures.
```
Comment (konradh) [inline packages/RichTextEditing-Core.package/DocumentReadStream.class/instance/endOfStructure.st]: Create a separate method for this (`structureDifferenceBetween: firstPosition and: secondPosition` or `structureDifferenceToNext`?) as it is used in `endOfStucture` and `beginningOfStructure` and maybe somewhere else in the future.

File: packages/RichTextEditing-Core.package/DocumentReadStream.class/instance/on..st [hunk 19]
```diff
@@ -0,0 +1,6 @@
+initialize
+on: aDocument
+
+	collection := aDocument text.
+	readLimit := collection size.
```
Comment (konradh) [inline packages/RichTextEditing-Core.package/DocumentReadStream.class/instance/on..st:5]: Maybe set readLimit in the accessor for collection. Or is it necessary to set one but not the other?

File: packages/RichTextEditing-Core.package/DocumentReadStream.class/instance/lineBreak.st [hunk 20]
```diff
@@ -0,0 +1,6 @@
+stop conditions
+lineBreak
+
+	^ ((self collection at: self position) = Character lf)
```
Comment (konradh) [inline packages/RichTextEditing-Core.package/DocumentReadStream.class/instance/lineBreak.st]: Create a method for this (`currentCharacterIs: aCharacter`).

File: packages/RichTextEditing-Core.package/DocumentReadStream.class/instance/readLimit..st [hunk 21]
```diff
@@ -0,0 +1,4 @@
+stop conditions
```
Comment (konradh) [inline packages/RichTextEditing-Core.package/DocumentReadStream.class/instance/readLimit..st:1]: accessing

File: packages/RichTextEditing-Core.package/DocumentReadStream.class/instance/on..st (hunk 19 above)
Comment (oleschl) [reply packages/RichTextEditing-Core.package/DocumentReadStream.class/instance/on..st:5]: I think you should not be able to change the text after opening a stream. So I would keep the way it is and also suggest to delete the setters for text and readLimit.

File: packages/RichTextEditing-Core.package/DocumentReadStream.class/instance/lineBreak.st (hunk 20 above)
Comment (oleschl) [reply packages/RichTextEditing-Core.package/DocumentReadStream.class/instance/lineBreak.st]: What do you think about the method ```currentCharacter``` to make it more flexible?

File: packages/RichTextEditing-Core.package/DocumentReadStream.class/instance/paragraphCharacter.st [hunk 22]
```diff
@@ -0,0 +1,4 @@
+constants
+paragraphCharacter
+
+	^Character value: 182
```
Comment (j-hellenberg) [inline packages/RichTextEditing-Core.package/DocumentReadStream.class/instance/paragraphCharacter.st]: Add space

File: packages/RichTextEditing-Tests.package/RichTextDocumentReadStreamTests.class/instance/isEndOfStructure..st [hunk 23]
```diff
@@ -0,0 +1,6 @@
+utilities
+isEndOfStructure: listElement
+
+	^ listElement first = #endOfStructure
+
+
```
Comment (j-hellenberg) [inline packages/RichTextEditing-Tests.package/RichTextDocumentReadStreamTests.class/instance/isEndOfStructure..st]: Too many newlines

File: packages/RichTextEditing-Tests.package/RichTextDocumentReadStreamTests.class/instance/textWithAttributes.st [hunk 24]
```diff
@@ -0,0 +1,8 @@
+constants
+textWithAttributes
+
+	| text |
+	text := ('hello', self paragraphCharacter asString, 'world') asText.
+	text := text addAttribute: (self boldTextStructure) from: 1 to: 6.
+	text := text addAttribute: (self italicTextStructure) from: 7 to: 11.
+	^text
```
Comment (j-hellenberg) [inline packages/RichTextEditing-Tests.package/RichTextDocumentReadStreamTests.class/instance/textWithAttributes.st]: space

File: packages/RichTextEditing-Tests.package/RichTextDocumentReadStreamTests.class/instance/textWithAttributes.st [hunk 25]
```diff
@@ -0,0 +1,8 @@
+constants
+textWithAttributes
+
+	| text |
+	text := ('hello', self paragraphCharacter asString, 'world') asText.
+	text := text addAttribute: (self boldTextStructure) from: 1 to: 6.
+	text := text addAttribute: (self italicTextStructure) from: 7 to: 11.
```
Comment (j-hellenberg) [inline packages/RichTextEditing-Tests.package/RichTextDocumentReadStreamTests.class/instance/textWithAttributes.st]: Not sure if the parenthesis are needed here

File: packages/RichTextEditing-Tests.package/RichTextEditorTestCase.class/instance/boldTextStructure.st [hunk 26]
```diff
@@ -0,0 +1,4 @@
+constants
+boldTextStructure
```
Comment (j-hellenberg) [inline packages/RichTextEditing-Tests.package/RichTextEditorTestCase.class/instance/boldTextStructure.st:2]: Isn't this already defined in TextStructure?

Comment (frcroth) [reply packages/RichTextEditing-Tests.package/RichTextEditorTestCase.class/instance/boldTextStructure.st:2]: convenience/ abstraction

**Pull request merged**